File: src/gx/commands/oops.py
"""gx oops — Quick-fix the last commit."""

from __future__ import annotations

import os
from typing import List, Optional

import typer

from gx.utils.display import (
    confirm_action,
    console,
    print_dry_run,
    print_error,
    print_info,
    print_success,
    print_warning,
)
from gx.utils.git import (
    GitError,
    ensure_git_repo,
    get_last_commit,
    is_commit_pushed,
    run_git,
)


def oops(
    message: Optional[str] = typer.Option(None, "-m", "--message", help="New commit message."),
    add: Optional[List[str]] = typer.Option(None, "--add", help="File(s) to add to the last commit."),
    dry_run: bool = typer.Option(False, "--dry-run", help="Show what would change."),
    force: bool = typer.Option(False, "--force", help="Allow amending even if already pushed."),
) -> None:
    """Quick-fix the last commit \u2014 amend message, add forgotten files, or both."""
    try:
        ensure_git_repo()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    commit = get_last_commit()
    if not commit:
        print_error("No commits yet.")
        raise typer.Exit(1)

    # Safety: check if pushed
    if is_commit_pushed() and not force:
        print_error(
            "The last commit has already been pushed to remote.\n"
            "  Amending it would rewrite shared history.\n"
            "  Use --force to override (dangerous)."
        )
        raise typer.Exit(1)

    if is_commit_pushed() and force:
        print_warning("The last commit has been pushed. Amending will require a force push.")

    # Validate --add files
    if add:
        for filepath in add:
            if filepath != "." and not os.path.exists(filepath):
                print_error(f"File not found: {filepath}")
                raise typer.Exit(1)

    # If neither message nor add specified, open editor for message
    if message is None and not add:
        # Amend with editor
        console.print()
        console.print(f'[bold]Last commit:[/bold] "{commit["message"]}" ({commit["short_hash"]})')
        console.print()

        if dry_run:
            print_dry_run(["Would open editor to amend commit message."])
            return

        if not confirm_action("Open editor to amend commit message?"):
            print_info("Cancelled.")
            raise typer.Exit(0)

        try:
            import subprocess
            subprocess.run(["git", "commit", "--amend"], check=True)
            print_success("Commit message amended.")
        except (subprocess.CalledProcessError, FileNotFoundError) as e:
            print_error(f"Failed to amend: {e}")
            raise typer.Exit(1)
        return

    # Show what will happen
    console.print()
    console.print(f'[bold]Last commit:[/bold] "{commit["message"]}" ({commit["short_hash"]})')
    console.print()

    actions = []
    dry_actions = []

    if add:
        # Check for actual changes in files
        files_to_add = []
        for filepath in add:
            if filepath == ".":
                files_to_add.append(".")
                actions.append("Adding all unstaged changes to last commit")
                dry_actions.append("Would add all unstaged changes to last commit")
            else:
                # Check if file has changes
                diff = run_git(["diff", "--name-only", filepath], check=False)
                staged = run_git(["diff", "--cached", "--name-only", filepath], check=False)
                status = run_git(["status", "--porcelain", filepath], check=False)
                if not diff and not staged and not status:
                    print_info(f"No changes in {filepath} — skipping.")
                    continue
                files_to_add.append(filepath)
                actions.append(f"  + {filepath}")
                dry_actions.append(f"Would add {filepath} to last commit")

        if not files_to_add:
            print_info("No files with changes to add.")
            raise typer.Exit(0)

        console.print("  [bold]Adding to last commit:[/bold]")
        for a in actions:
            console.print(f"    {a}")
        console.print()

    if message:
        console.print("  [bold]Amending message:[/bold]")
        console.print(f'    Before: "{commit["message"]}"')
        console.print(f'    After:  "{message}"')
        console.print()
        dry_actions.append(f'Would change message from "{commit["message"]}" to "{message}"')

    if not message and add:
        console.print(f'  Commit message stays: "{commit["message"]}"')
        console.print()

    if dry_run:
        print_dry_run(dry_actions)
        return

    if not confirm_action("Proceed?"):
        print_info("Cancelled.")
        raise typer.Exit(0)

    # Execute
    try:
        if add:
            for filepath in add:
                run_git(["add", filepath])

        amend_args = ["commit", "--amend"]
        if message:
            amend_args.extend(["-m", message])
        else:
            amend_args.append("--no-edit")
        run_git(amend_args)
    except GitError as e:
        print_error(f"Failed: {e}")
        raise typer.Exit(1)

    console.print()
    if message and add:
        print_success("File added and commit message amended.")
    elif message:
        print_success("Commit message amended.")
    elif add:
        print_success("File added to last commit.")

File: cmd/oops.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"

	"github.com/mubbie/gx-cli/internal/git"
	"github.com/mubbie/gx-cli/internal/ui"
	"github.com/spf13/cobra"
)

func init() {
	cmd := &cobra.Command{
		Use:   "oops",
		Short: "Quick-fix the last commit. Amend message, add forgotten files, or both.",
		RunE:  runOops,
	}
	cmd.Flags().StringP("message", "m", "", "New commit message")
	cmd.Flags().StringSlice("add", nil, "File(s) to add to the last commit")
	cmd.Flags().Bool("dry-run", false, "Show what would change")
	cmd.Flags().Bool("force", false, "Allow amending even if already pushed")
	rootCmd.AddCommand(cmd)
}

func runOops(cmd *cobra.Command, args []string) error {
	if err := git.EnsureRepo(); err != nil {
		ui.PrintError(err.Error())
		return nil
	}

	_, shortHash, lastMsg, _, _ := git.LastCommit()
	if shortHash == "" {
		ui.PrintError("No commits yet.")
		return nil
	}

	force, _ := cmd.Flags().GetBool("force")
	if git.IsCommitPushed() && !force {
		ui.PrintError("The last commit has already been pushed to remote.\n  Amending it would rewrite shared history.\n  Use --force to override (dangerous).")
		return nil
	}

	message, _ := cmd.Flags().GetString("message")
	addFiles, _ := cmd.Flags().GetStringSlice("add")
	dryRun, _ := cmd.Flags().GetBool("dry-run")

	// No flags: open editor
	if message == "" && len(addFiles) == 0 {
		if dryRun {
			ui.PrintDryRun([]string{"Would open editor to amend commit message."})
			return nil
		}
		fmt.Printf("Last commit: \"%s\" (%s)\n", lastMsg, shortHash)
		if !ui.Confirm("Open editor to amend commit message?") {
			ui.PrintInfo("Cancelled.")
			return nil
		}
		c := exec.Command("git", "commit", "--amend")
		c.Stdin = os.Stdin
		c.Stdout = os.Stdout
		c.Stderr = os.Stderr
		if err := c.Run(); err != nil {
			ui.PrintError(fmt.Sprintf("Failed to amend: %s", err))
			return nil
		}
		ui.PrintSuccess("Commit message amended.")
		return nil
	}

	fmt.Printf("\nLast commit: \"%s\" (%s)\n\n", lastMsg, shortHash)

	var dryActions []string

	if len(addFiles) > 0 {
		fmt.Println("  Adding to last commit:")
		for _, f := range addFiles {
			if f != "." {
				if _, err := os.Stat(f); os.IsNotExist(err) {
					ui.PrintError(fmt.Sprintf("File not found: %s", f))
					return nil
				}
			}
			fmt.Printf("    + %s\n", f)
			dryActions = append(dryActions, fmt.Sprintf("Would add %s to last commit", f))
		}
		fmt.Println()
	}

	if message != "" {
		fmt.Printf("  Amending message:\n    Before: \"%s\"\n    After:  \"%s\"\n\n", lastMsg, message)
		dryActions = append(dryActions, fmt.Sprintf("Would change message to \"%s\"", message))
	}

	if dryRun {
		ui.PrintDryRun(dryActions)
		return nil
	}

	if !ui.Confirm("Proceed?") {
		ui.PrintInfo("Cancelled.")
		return nil
	}

	for _, f := range addFiles {
		if _, err := git.Run("add", f); err != nil {
			ui.PrintError(fmt.Sprintf("Failed to add %s: %s", f, err))
			return nil
		}
	}

	amendArgs := []string{"commit", "--amend"}
	if message != "" {
		amendArgs = append(amendArgs, "-m", message)
	} else {
		amendArgs = append(amendArgs, "--no-edit")
	}
	if _, err := git.Run(amendArgs...); err != nil {
		ui.PrintError(fmt.Sprintf("Failed: %s", err))
		return nil
	}

	fmt.Println()
	if message != "" && len(addFiles) > 0 {
		ui.PrintSuccess("File added and commit message amended.")
	} else if message != "" {
		ui.PrintSuccess("Commit message amended.")
	} else {
		ui.PrintSuccess("File added to last commit.")
	}
	return nil
}
